resource.go

```diff
@@ -60,13 +82,79 @@ func getResourceField(v interface{}) (f *structs.Field, err error) {
 	return
 }
 
+func getIdFromMap(m map[string]interface{}) (ret int, err error) {
+	if val, ok := m["id"].(int); !ok {
+		ret = 0
+	} else {
+		ret = val
+	}
+	return
+}
+
 // Retrieve Id field from interface.
 func resourceId(v interface{}) (ret int, err error) {
 	var f *structs.Field
+	invalidTypeErr := errors.New("Not a struct, map, or pointer to either")
```

Comment:
Why declare this as a separate variable? I think it'd be fine set directly on L114 as `err := errors.New(...)`.